mm0-rs/src/elab/lisp/debug.rs

```diff
@@ -0,0 +1,1465 @@
+//! Debug formatting for mm0-rs items. Follows the API set out by the `print` and `pretty`
+//! modules. 
+use std::collections::HashMap;
+use std::cell::RefCell;
+use std::{fmt::{self}};
+use shoebill::{ Printer, Doclike, DocPtr };
+use shoebill::ron::{ RonStruct, RonSequence, RonTuple, RonOption, RonResult };
+use super::{LispVal, LispKind, print::FormatEnv,
+  super::{
+    environment::{DeclKey, NotaInfo, AtomData, AtomID, SortID, TermID, ThmID, Term, Thm, Type, Sort, Expr }, }};
+use crate::util::{ FileSpan, FileRef, Span };
+
+/// Debug counterpart to the `Print` struct`.
+pub struct DebugPrint<'a, D : ?Sized> {
+  fe : FormatEnv<'a>,
+  e : &'a D
+}
+
+/// DebugPrint items can be put in formatters using `{:?}`
+impl<'a, D: EnvDebug + ?Sized> fmt::Debug for DebugPrint<'a, D> {
+  fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result { self.e.dbg(self.fe, f) }
+}
+
+impl<'a> FormatEnv<'a> {
+  /// The debug counterpart to `FormatEnv::pretty()`
+  pub fn debug<T>(self, f : impl for<'b> FnOnce(FormatEnv<'b>, &'b mut Printer<'b>) -> T) -> T {
+    f(self, &mut Printer::new())
+  }
+
+  /// The debug counterpart to `FormatEnv::to()`
+  pub fn to_debug<D : ?Sized>(self, e : &'a D) -> DebugPrint<'a, D> {
+    DebugPrint {fe : self, e}
+  }
+}
+
+impl EnvDebug for NotaInfo {
```

Comment:
This is all very boilerplate, and it looks like an additional maintenance burden as these structures change. I can see that some of the basic types require custom implementations, but I would suggest finding a way to do this with a macro.